executor/src/witgen/machines/machine_extractor.rs

```diff
@@ -64,18 +67,44 @@ pub fn split_out_machines<'a, T: FieldElement>(
     let mut base_identities = identities.clone();
     let mut extracted_prover_functions = HashSet::new();
     let mut id_counter = 0;
-    for id in &identities {
-        // Extract all witness columns in the RHS of the lookup.
-        let lookup_witnesses = match id {
-            Identity::Lookup(LookupIdentity { right, .. })
-            | Identity::PhantomLookup(PhantomLookupIdentity { right, .. })
-            | Identity::Permutation(PermutationIdentity { right, .. })
-            | Identity::PhantomPermutation(PhantomPermutationIdentity { right, .. }) => {
-                &refs_in_selected_expressions(right) & (&remaining_witnesses)
+
+    let all_connections = identities
+        .iter()
+        .filter_map(|i| Connection::try_from(*i).ok())
+        .collect::<Vec<_>>();
+
+    let mut fixed_lookup_connections = BTreeMap::new();
+
+    for connection in &all_connections {
+        // If the RHS only consists of fixed columns, record the connection and continue.
+        if connection.is_lookup()
+            && connection.right.selector.is_one()
+            && connection.right.expressions.iter().all(|e| {
+                try_to_simple_poly_ref(e)
+                    .map(|poly| poly.poly_id.ptype == PolynomialType::Constant)
+                    .unwrap_or(false)
+            })
+            && !connection.right.expressions.is_empty()
+        {
+            assert!(fixed_lookup_connections
+                .insert(connection.id, *connection)
+                .is_none());
+            if let Some(multiplicity) = connection.multiplicity_column {
+                remaining_witnesses.remove(&multiplicity);
             }
-            _ => Default::default(),
+            continue;
+        }
+
+        // Extract all witness columns in the RHS of the lookup.
+        let lookup_witnesses = {
+            let callee_columns = refs_in_selected_expressions(connection.right)
```

Comment:
Why not introduce a function `refs_in_connection_rhs` that includes the multiplicity?

Reply:
Done! I think this actually fixed a bug: If there are multiple phantom lookups with distinct multiplicities, we now include all multiplicities in the machine witness.